docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,316 @@
+
+
+## Proposal
+
+A Hybrid design
+
+Evidence has the following simple interface:
+
+```go
+type Evidence interface {  //proposed
+  Height() int64                                     // height of the equivocation
+  Bytes() []byte                                     // bytes which comprise the evidence
+  Hash() []byte                                      // hash of the evidence
+  ValidateBasic() error
+  String() string
+}
+```
+
+We have two concrete types of evidence that fulfil this interface
+
+```go
+type LightClientAttackEvidence struct {
+  ConflictingBlock *LightBlock
+  CommonHeight int64
+}
+```
+where the `Hash()` is the hash of the header and commonHeight
+
+```go
+type DuplicateVoteEvidence {
+  VoteA *Vote
+  VoteB *Vote
+}
+```
+where the `Hash()` is the hash of the two votes
+
+The first is generated in the light client, the second in consensus. Both are sent to the evidence pool through `AddEvidence(ev Evidence) error` where it first runs `Has(ev Evidence)` to check if it has already received it (by comparing hashes) and then  `Verify(ev Evidence) error` before adding it to the database. There are two databases: one
+for pending evidence that is not yet committed and another of the committed evidence (on the chain)
+
+
+`Verify()` does the following:
+
+- Use the hash to see if we already have this evidence in our committed database.
+
+- Use the height to check if the evidence hasn't expired.
+
+- If it hasn't expired use the height to find the block header and confirm that the time hasn't also expired.
+
+- Then proceed with switch statement for each of the two evidence:
+
+For `DuplicateVote`:
+
+- Check that height, round, type and validator address are the same
+
+- Check that the Block ID is different
+
+- Check the look up table for addresses to make sure there already isn't evidence against this validator
+
+- Fetch the validator set and confirm that the address is in the set at the height of the attack
+
+- Check that the chain ID and signature is valid.
+
+For `LightClientAttack`
+
+- Fetch the common signed header and val set from the common height and use skipping verification to verify the conflicting header
+
+- Fetch the trusted signed header at the same height as the conflicting header and compare with the conflicting header to work out which type of attack it is and in doing so return the malicious validators.
+
+  - If equivocation the validators that signed for the commits of both the trusted and signed header
+
+  - If lunatic the validators from the common val set that signed in the conflicting block
+
+  - If amnesia none (we can't know which validators are malicious). This also means that we don't
+  send amnesia evidence to the application
+
+- For each validator, check the look up table to make sure there already isn't evidence against this validator
```

Comment:
what if it's a newer evidence? old one might expire

Reply:
So my thinking on this is that we want to have this look up table as a way to stop malicious validators from spamming the network by creating thousands of evidence right? So if each evidence is actually valid and we always check and keep the newest one then this is not going solve our problem as we will keep having to validate and update our lookup table and then try gossip every new evidence that comes in. In all honesty, I would like to think that this evidence expiration will never really be a problem unless of course someone is intentionally censoring the network (I'm expecting there to be a large buffer between trusting period and expiration period).

However, if we wanted to try do both we could have a check to see if the existing header is in the later half of that period  and if so then we can look to swap it out with the newer evidence

Reply:
yeah, let's implement a simpler version first